src/workerd/api/streams/readable-source.c++

```diff
@@ -529,40 +462,134 @@ class ReadableStreamSourceImpl: public ReadableStreamSource {
-        KJ_ASSERT(minBytes <= buffer.size());
-        kj::AsyncInputStream& stream =
-            option.identityEncoding ? ensureIdentityEncoding(inner) : *inner;
-        try {
-          // The read() method on AsyncInputStream will throw an exception on short reads.
-          auto amount = co_await stream.tryRead(buffer.begin(), minBytes, buffer.size());
-          if (amount < minBytes) {
-            setClosed();
-          }
-          co_return amount;
-        } catch (...) {
-          auto exception = kj::getCaughtExceptionAsKj();
-          if (exception.getType() == kj::Exception::Type::DISCONNECTED) {
-            // Treat disconnect as EOF.
-            setClosed();
-            co_return 0;
+
+      // Set the write buffer to the one we just filled.
+      auto writeBuf = buffer[current];
+
+      // Then switch to the other buffer and start the next read.
+      current = 1 - current;
+
+      // Before we perform the next read, let's adapt minBytes based on stream behavior
+      // we have observed on the previous read.
+      if (iterationCount <= 3 || iterationCount % 10 == 0) {
+        if (amount == bufferSize) {
+          // Stream is filling buffer completely... Use smaller minBytes to
+          // increase responsiveness, should produce more reads with less data.
+          minBytes = kj::max(bufferSize >> 2, kj::min(DEFAULT_BUFFER_SIZE, bufferSize >> 1));
+        } else {
+          // Stream is moving slower, increase minBytes to try to get larger chunks.
+          minBytes = (bufferSize >> 2) + (bufferSize >> 1);  // 75%
+        }
+      }
+
+      KJ_IF_SOME(remaining, maybeRemaining) {
+        if (amount > remaining) {
+          // The stream lied about its length. Ignore further length tracking.
+          maybeRemaining = kj::none;
+        } else {
+          // Otherwise, set minBytes to whatever is expected to remain.
+          remaining -= amount;
+          maybeRemaining = remaining;
+          if (remaining < minBytes && remaining > 0) {
+            minBytes = remaining;
           }
-          setErrored(kj::cp(exception));
-          kj::throwFatalException(kj::mv(exception));
         }
       }
+
+      // Start our next read operation.
+      readPromise = readImpl(*stream, buffer[current], minBytes);
+
+      // Write out the chunk we just read in parallel with the next read.
+      // If the write fails, the exception will propagate and cancel the pump,
+      // including the read operation. If the read fails, it will be picked
+      // up at the start of the next loop iteration.
+      co_await output.write(writeBuf.first(amount));
+    }
+
+    if (end) co_await output.end();
+  }
+
+  static kj::Promise<size_t> readImpl(
+      kj::AsyncInputStream& inner, kj::ArrayPtr<kj::byte> buffer, size_t minBytes) {
+    KJ_ASSERT(minBytes <= buffer.size());
+    try {
+      // The read() method on AsyncInputStream will throw an exception on short reads.
```

Comment:
But `tryRead()` does not, which is what you're using below

Reply:
Expanded the comment.

Reply:
I see the expanded comment, but the code doesn't quite make sense to me now. Why is the `tryRead()` wrapped in a try-catch? Converting network disconnections into clean EOFs isn't necessarily correct, and could hide data loss.